File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
let fs = require("fs");
let _ = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

let [N, ...arr] = _;

let [n, v] = N.split(" ").map(Number);
arr = arr.map((v) => v.split(" ").map(Number));

function countConnectedComponents(N, edges) {
  const graph = new Array(N).fill(null).map(() => []);
  const visited = new Array(N).fill(false);

  function dfs(node) {
    visited[node] = true;
    for (let neighbor of graph[node]) {
      if (!visited[neighbor]) {
        dfs(neighbor);
      }
    }
  }

  for (let edge of edges) {
    const [u, v] = edge;
    graph[u - 1].push(v - 1);
    graph[v - 1].push(u - 1);
  }

  let components = 0;
  for (let i = 0; i < N; i++) {
    if (!visited[i]) {
      dfs(i);
      components++;
    }
  }

  return components;
}

console.log(countConnectedComponents(n, arr));

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
import collections
import math
from collections import deque
import copy
import bisect
import itertools
import heapq

#sys.stdin = open("input.txt", "r")
input = sys.stdin.readline

N,M = map(int,input().split())
graph = [[] for _ in range(N+1)]
for _ in range(M):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)

visited = [0 for _ in range(N+1)]

def dfs(node):
    visited[node] = 1
    for nextNode in graph[node]:
        if visited[nextNode] == 1:
            continue
        visited[nextNode] = 1
        dfs(nextNode)

def bfs(x):
    queue = deque()
    queue.append(x)
    visited[x] = 1
    while queue:
        node = queue.popleft()
        for nextNode in graph[node]:
            if visited[nextNode] == 1:
                continue
            visited[nextNode] = 1
            queue.append(nextNode)
ans = 0
for i in range(1,N+1):
    if visited[i] == 0:
        ans += 1
        bfs(i)
        
print(ans)